src/python/pants/backend/codegen/protobuf/scala/ScalaPBShim.scala

```diff
@@ -0,0 +1,123 @@
+import scalapb.ScalaPbCodeGenerator
+
+
+// Derived in part from ScalaPBC under Apache License v2.0. Introduced errors are mine.
+
+case class Config(
+  protocPath: Option[String] = None,
+  throwException: Boolean = false,
+  args: Seq[String] = Seq.empty,
+  namedGenerators: Seq[(String, ProtocCodeGenerator)] = Seq("scala" -> ScalaPbCodeGenerator),
+  executableArtifacts: Seq[String] = Seq.empty,
+  jvmPlugins: Seq[(String, String)] = Seq.empty
+)
+
+class ScalaPbcException(msg: String) extends RuntimeException(msg)
+
+object ScalaPBShim {
+  private val ProtocPathArgument = "--protoc="
+  private val JvmPluginArgument = "--jvm-plugin="
+
+  def processArgs(args: Array[String]): Config = {
+    case class State(cfg: Config, passThrough: Boolean)
+
+    args
+      .foldLeft(State(Config(), false)) { case (state, item) =>
+        (state.passThrough, item) match {
+          case (false, "--")      => state.copy(passThrough = true)
+          case (false, "--throw") => state.copy(cfg = state.cfg.copy(throwException = true))
+          case (false, p) if p.startsWith(ProtocPathArgument) =>
+            state.copy(
+              cfg = state.cfg
+                .copy(protocPath = Some(p.substring(ProtocPathArgument.length)))
+            )
+          case (false, p) if p.startsWith(JvmPluginArgument) =>
+            val Array(genName, classpath) = p.substring(JvmPluginArgument.length).split('=')
+            state.copy(
+              cfg = state.cfg.copy(jvmPlugins = state.cfg.jvmPlugins :+ (genName -> classpath))
+            )
+          case (_, other) =>
+            state.copy(passThrough = true, cfg = state.cfg.copy(args = state.cfg.args :+ other))
+        }
+      }
+      .cfg
+  }
+
+  def findMainClass(f: File): Either[String, String] = {
+    val jin = new JarInputStream(new FileInputStream(f))
+    try {
+      val manifest = jin.getManifest()
+      Option(manifest.getMainAttributes().getValue("Main-Class"))
+        .toRight("Could not find main class for plugin")
+        .map(_ + "$")
+    } finally {
+      jin.close()
+    }
+  }
+
+  private[scalapb] def runProtoc(config: Config): Int = {
+    if (
+      config.namedGenerators
+        .map(_._1)
+        .toSet
+        .intersect(config.jvmPlugins.map(_._1).toSet)
+        .nonEmpty
+    ) {
+      throw new RuntimeException(
+        s"Plugin name conflict with $JvmPluginArgument"
+      )
+    }
+
+    def fatalError(err: String): Nothing = {
+      if (config.throwException) {
+        throw new ScalaPbcException(s"Error: $err")
+      } else {
+        System.err.println(err)
+        sys.exit(1)
+      }
+    }
+
+    val jvmGenerators = config.jvmPlugins.map({ case (name, classpath) =>
+      val files = classpath.split(':').map(f => new File(f))
+      val urls = files.map(_.toURI().toURL()).toArray
+      val loader = new URLClassLoader(urls, null)
+      val mainClass = findMainClass(files.last) match {
```

Comment:
This code assumes that the last entry is the jar of the protoc plugin. This is probably too reliant on impl of `.classpath_entries` on the Python side.

Reply:
Probably worth an error enrichment on the `Left` case, but fine for now.